File: 1298-maximum-candies-you-can-get-from-boxes/1298-maximum-candies-you-can-get-from-boxes.cs
public class Solution {
    public int MaxCandies(int[] status, int[] candies, int[][] keys, int[][] containedBoxes, int[] initialBoxes) {
        int n = candies.Length, maxCandies = 0;
        bool[] boxFound = new bool[n], keyFound = new bool[n];

        var queue = new Queue<int>();
        foreach(var box in initialBoxes){
            boxFound[box] = true;
            if(status[box] == 1) queue.Enqueue(box);
        }

        while(queue.Count > 0){
            var box = queue.Dequeue();
            if(status[box] == -1) continue; // already explored

            // capture candies
            maxCandies += candies[box];
            status[box] = -1;   // mark visited

            // explore new boxes found
            foreach(var cBox in containedBoxes[box]){
                boxFound[cBox] = true;
                if(status[cBox] == 1 || (status[cBox] == 0 && keyFound[cBox])) queue.Enqueue(cBox);
            }

            // explore new keys found
            foreach(var kBox in keys[box]){
                keyFound[kBox] = true;
                if(status[kBox] != -1 && boxFound[kBox]) queue.Enqueue(kBox);
            }
        }

        return maxCandies;
    }
}

File: 1298-maximum-candies-you-can-get-from-boxes/1298-maximum-candies-you-can-get-from-boxes.java
class Solution {
    public int maxCandies(int[] status, int[] candies, int[][] keys, int[][] containedBoxes, int[] initialBoxes) {
        int n = candies.length, maxCandies = 0;
        boolean[] boxFound = new boolean[n], keyFound = new boolean[n];

        Queue<Integer> queue = new LinkedList<>();
        for(var box : initialBoxes){
            boxFound[box] = true;
            if(status[box] == 1) queue.offer(box);
        }

        while(!queue.isEmpty()){
            var box = queue.poll();
            if(status[box] == -1) continue; // already explored

            // capture candies
            maxCandies += candies[box];
            status[box] = -1;   // mark visited

            // explore new boxes found
            for(var cBox : containedBoxes[box]){
                boxFound[cBox] = true;
                if(status[cBox] == 1 || (status[cBox] == 0 && keyFound[cBox])) queue.offer(cBox);
            }

            // explore new keys found
            for(var kBox : keys[box]){
                keyFound[kBox] = true;
                if(status[kBox] != -1 && boxFound[kBox]) queue.offer(kBox);
            }
        }

        return maxCandies;
    }
}